### DISCUSSION TITLE
Use of `mixin` with `spread` gives unexpected result

### DISCUSSION THREAD

**Discussion by dshatokhin** [apple/pkl#562] [Q&A] [answered]
Hi, everyone!

I'm trying to deeply merge 2 dynamic objects and found out about `Mixin` from discussion #505 .
It's working as expected while using suggested approach:

```pkl
// Working example

hidden defaultParams {
  options {
    enabled = false
    someOpt = "someValue"
  }
}

hidden overridesLiteral: Mixin<Dynamic> = new {
  options {
    enabled = true
  }
}

merged = defaultParams |> overridesSpread
```

Gives expected output:

```pkl
merged {
  options {
    enabled = true
    someOpt = "someValue"
  }
}
```

However, when using `spread` for `Mixin` it's producing different result:

```pkl
// Spread example

hidden defaultParams {
  options {
    enabled = false
    someOpt = "someValue"
  }
}

hidden customParams {
  options {
    enabled = true
  }
}

hidden overridesSpread: Mixin<Dynamic> = new {
  ...customParams
}

merged = defaultParams |> overridesSpread
```

Gives different output:

```pkl
merged {
  options {
    enabled = true
  }
}
```

What am I missing here and could it even be archived with use of a `spread`?

**Comment by HT154** [comment]
This is an extension of the case that came up in #505: the spread operator performs a property/member-wise _assignment_. In contrast, applying a mixin performs a property/element-wise _amend_.

Accordingly, in your example, the spread usage:
```

hidden overridesSpread: Mixin<Dynamic> = new {
  ...customParams
}
```
Is "de-sugared" to:
```
hidden overridesSpread: Mixin<Dynamic> = new {
  options = customParams.options
}
```

And here it's clearer to see that the resulting value for `merged.options` is exactly what was specified in `customParams` and not amended by it.

It's a little tricky to determine what you're going for overall here, but if I'm inferring correctly, it seems like `customParams` should itself be defined as a mixin from the start here to achieve the desired results.

**Comment by dshatokhin** [comment]
Thanks for the explanation, now I see what is going on.

While reducing my problem to provided examples I completely missed the important context - I'm reading `customParams` from a `yaml` file:

```yaml
# customParams.yaml
options:
  enabled: true
```

And then the full example would be as such:

```pkl
import "pkl:yaml"

hidden defaultParams {
  options {
    enabled = false
    someOpt = "someValue"
  }
}

hidden customParams = new yaml.Parser {}.parse(read("file:customParams.yaml"))

hidden overridesSpread: Mixin<Dynamic> = new {
  ...customParams
}

merged = defaultParams |> overridesSpread
```

```pkl
merged {
  options {
    enabled = true
  }
}
```

I've tried 

`overrides: Mixin = new yaml.Parser {}.parse(read("file:customParams.yaml"))`

but got an error:

`Expected value of type 'Function1', but got type 'Dynamic'.`

Is it possible to parse the file directly to `Mixin` typed variable?

**Reply by holzensp** [reply]
Do you have Pkl type information for the intended parse result? As in, do you know what properties to expect and what types they have and do you have the module definition to match?

If you do, you can use [the `deepToTyped` package](https://pkl-lang.org/package-docs/pkg.pkl-lang.org/pkl-pantry/pkl.experimental.deepToTyped/current/index.html) to one-shot check the `yaml` being read. After that, Pkl would give you at least errors, telling you that you're not getting what you think you are. They could use a little more explanation, here and there (if ever you find errors uninformative, please flag it).

If you really want to use the `yaml` in the same way Pkl does amending (in other words, if property `foo: bar` is in the `yaml`, then the `Mixin` equivalent would be `foo = "bar"`), you can probably do something like
```
overrides: Mixin = (it) -> (it.toMap() + new yaml.Parser {}.parse(read("file:customParams.yaml")).toMap()).toTyped(NotSure)
```

Edit: Note that `+` on `Map`s has keys in the right-hand-side argument overwrite those in the left. Also good to know that a `Mixin<T>` can be read as a `(T) -> T` function and that's why the `toTyped(NotSure)`. You'll likely get away with it, because none of the types are constrained. That does mean that unexpected things happen quietly.

This does raise all sorts of questions about what you want the recursive behaviour to be. Should a top-level name-match completely overwrite, or should it recursively mix in? Also, have you heard of [the Norway Problem](https://hitchdev.com/strictyaml/why/implicit-typing-removed/)? Here be dragons; before you know it, you're re-implementing Pkl in Pkl. I have a feeling there may be a nested `for` generator that could express what you want to do.

**Reply by holzensp** [reply]
Considering you call it `customParams`, you may not be asking for a `Mixin` because you want to do it for a bunch of files, but just because you want the value of a property in the yaml file to "win" over some default. If that is the case you can do something like this;
```
import "pkl:yaml"

local yamlContents = new yaml.Parser {}.parse(read("file:customParams.yaml")) as Dynamic

foo: String = yamlContents.getPropertyOrNull("foo") ?? "foo value in case no foo in yaml"
```

That does require you to name all the properties you want to keep.

**Reply by dshatokhin** [reply]
Thanks for insights, it helps a lot to break a tunnel vision I acquired while working on the problem.
And now I know the proper name for the problem with `no` in `yaml` files.

I have a `class` defined as well as default values for it, so I am expecting the same structure in the `customParams.yaml`.
The `deepToTyped` package used a bit later in my code, after the combining `default` and `custom` parameters.

Right now, I'm trying to implement something like `helm` where I have default values for generating kubernetes manifests and also want to have an ability to pass overrides.

I'm doing it for the sake of learning and showcasing it on a meetup.

So, the idea is to do it as simple as possible and I started with `overrides = defaultParams.toMap() + customParams.toMap()` where my problem first occurred. My approach is to find the easiest and most straightforward solution before digging deeper.

Basically, I defined a default for generating kubernetes manifest, and seeking for a way to override only some of them in a nested object:

```pkl
defaultValues = new Values {
  livenessProbe = new {
    httpGet {
      path = "/"
      port = "http"
    }
  }
}
```

And from `customParams.yaml` I want only the absolute minimum to update the defaults:

```yaml
livenessProbe:
  httpGet:
    path: /live
```

I might be looking at the problem from the wrong angle and I think I could just rethink it or leave it as is and add a mention to my list of `known limitations`. Or maybe nested `for` could do the trick.

**Reply by holzensp** [reply]
Quick question; must the `overrides` be a YAML file, or could it be a Pkl file?

**Reply by holzensp** [reply, marked as answer]
I ask, because you're defining Pkl's _amending_ quite precisely there. It sounds like you want a `MyParams.pkl` (upper case, because a template with defaults and type definitions, meant to be amended) with something like:
```
class HttpRequest {
  hidden type: *"GET"|"POST"|"PUT" //...
  path: String(startsWith("/")) = "/"
  port: *"http"|"https"|Int(isBetween(1,65535))
}

class LivenessProbe {
  httpGet: HttpRequest(type == "GET")
}

livenessProbe: LivenessProbe

output {
  renderer = new YamlRenderer {}
}
```
and in a subdirectory, somewhere, you might have `myParams.pkl` (lower case, because amending a template);
```
amends ".../MyParams.pkl"

path = "/live"
```

Now when you `pkl eval myParams.pkl`, you get
```
livenessProbe:
  httpGet:
    path: /live
    port: "http"
```

**Reply by dshatokhin** [reply]
Well, I would prefer `yaml` instead of `pkl` but the point of what I'm doing is to showcase `pkl` as a templating language which aware of `kubernetes` objects (and it's working really well).

It seems `amends` exactly what I need and it's really straightforward, so I'll stick to it 👍

Once again, thanks for answering my questions and helping with examples 🥇
